include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -0,0 +1,371 @@
+ * underlying LAPACK implementation supports only float and double numbers, so
+ * only these two types are currently supported.
+ *
+ * @tparam size Compile-time array lenghts. By default at -1, which means that
+ * the run-time info stored in the matrices passed to the reinit()
+ * function is used.
+ *
+ * @author Martin Kronbichler, 2017
+ */
+template <int dim, typename Number, int size = -1>
+class TensorProductMatrix
+{
+public:
+  /**
+   * Constructor.
+   */
+  TensorProductMatrix();
+
+  /**
+   * Initalizes the matrix to the given mass matrix $M$ and derivative matrix
+   * $A$. Note that the current implementation requires $M$ to be symmetric
+   * and positive definite and $A$ to be symmetric.
+   */
+  void reinit (const FullMatrix<Number> &mass_matrix,
+               const FullMatrix<Number> &derivative_matrix);
+
+  /**
+   * Returns the number of rows of this matrix, given by the dim-th power of
+   * the size of the 1D matrices passed to the constructor.
+   */
+  unsigned int m() const;
+
+  /**
+   * Returns the number of columns of this matrix, given by the dim-th power
+   * of the size of the 1D matrices passed to the constructor.
+   */
+  unsigned int n() const;
+
+  /**
+   * Implements a matrix-vector product with the underlying matrix as
+   * described in the main documentation of this class.
+   */
+  void vmult (Vector<Number> &dst,
+              const Vector<Number> &src) const;
+
+  /**
+   * Implements a matrix-vector product with the underlying matrix as
+   * described in the main documentation of this class. Same as the other
+   * vmult function, but operating on plain pointers rather than a vector (no
+   * check of array bounds possible).
+   */
+  void vmult (Number *dst,
+              const Number *src) const;
+
+  /**
+   * Implements a matrix-vector product with the underlying matrix as
+   * described in the main documentation of this class.
+   */
+  void apply_inverse (Vector<Number> &dst,
+                      const Vector<Number> &src) const;
+
+  /**
+   * Implements a matrix-vector product with the underlying matrix as
+   * described in the main documentation of this class. Same as the other
+   * vmult function, but operating on plain pointers rather than a vector (no
+   * check of array bounds possible).
+   */
+  void apply_inverse (Number *dst,
+                      const Number *src) const;
+
+private:
+  /**
+   * A copy of the @p mass_matrix object passed to the reinit() method.
+   */
+  FullMatrix<Number> mass_matrix;
+
+  /**
+   * A copy of the @p derivative_matrix object passed to the reinit() method.
+   */
+  FullMatrix<Number> derivative_matrix;
+
+  /**
+   * A vector containing the generalized eigenvalues of A s = lambda B s.
+   */
+  AlignedVector<Number> eigenvalues;
+
+  /**
+   * The matrix containing the generalized eigenvectors.
+   */
+  Table<2,Number> eigenvectors;
+
+  /**
+   * An array for temporary data.
+   */
+  mutable AlignedVector<Number> tmp_array;
```

Comment:
If every thread creates its own TensorProductMatrix object, then the mutex doesn't do much harm: there is never contention for the mutex and acquiring it is cheap. Are you sure that mutex acquisition cost would be relevant at all compared to the rest of what the function does?

As for thread-local variables: this only works if nowhere in your code path between first and last access of the variable are you calling anything that might involve task switches or anything that might invoke the TBB. That's sometimes difficult to establish because it might happen in functions you call from here -- e.g., something like `vector=0`.